src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt

```diff
@@ -12,237 +12,277 @@ import kotlinx.coroutines.flow.*
-    object Connecting : ElectrumConnectionStatus
-    data class Connected(val version: ServerVersionResponse, val height: Int, val header: BlockHeader) : ElectrumConnectionStatus
+    fun toConnectionState(): Connection = when (this) {
+        is Closed -> Connection.CLOSED(this.reason)
+        Connecting -> Connection.ESTABLISHING
+        is Connected -> Connection.ESTABLISHED
+    }
 }
 
 class ElectrumClient(
-    socketBuilder: TcpSocket.Builder?,
-    scope: CoroutineScope,
-    private val loggerFactory: LoggerFactory
-) : CoroutineScope by scope, IElectrumClient {
+    private val scope: CoroutineScope,
+    private val loggerFactory: LoggerFactory,
+    private val pingInterval: Duration = 30.seconds,
+    private var rpcTimeout: Duration = 10.seconds
+) : IElectrumClient {
 
     private val logger = loggerFactory.newLogger(this::class)
 
-    var socketBuilder: TcpSocket.Builder? = socketBuilder
-        set(value) {
-            logger.debug { "swap socket builder=$value" }
-            field = value
-        }
-
     private val json = Json { ignoreUnknownKeys = true }
 
-    // new connection status
+    // connection status
     private val _connectionStatus = MutableStateFlow<ElectrumConnectionStatus>(ElectrumConnectionStatus.Closed(null))
     override val connectionStatus: StateFlow<ElectrumConnectionStatus> get() = _connectionStatus.asStateFlow()
 
-    // legacy connection status
-    private val _connectionState = MutableStateFlow<Connection>(Connection.CLOSED(null))
-    override val connectionState: StateFlow<Connection> get() = _connectionState.asStateFlow()
-
     // subscriptions notifications (headers, script_hashes, etc.)
     private val _notifications = MutableSharedFlow<ElectrumSubscriptionResponse>(replay = 0, extraBufferCapacity = 64, onBufferOverflow = BufferOverflow.SUSPEND)
     override val notifications: Flow<ElectrumSubscriptionResponse> get() = _notifications.asSharedFlow()
 
-    private sealed interface Action {
-        data class SendToServer(val request: Pair<ElectrumRequest, CompletableDeferred<ElectrumResponse>>) : Action
-        data class ProcessServerResponse(val response: Either<ElectrumSubscriptionResponse, JsonRPCResponse>) : Action
-        object Disconnect : Action
+    private sealed class Action {
+        data class SendToServer(val request: Pair<ElectrumRequest, CompletableDeferred<ElectrumResponse>>) : Action()
+        data class ProcessServerResponse(val response: Either<ElectrumSubscriptionResponse, JsonRPCResponse>) : Action()
     }
 
-    private var mailbox = Channel<Action>()
-
-    private val statusJob: Job
-    private var runJob: Job? = null
+    // This channel acts as a queue for messages sent/received to the electrum server.
+    // It lets us decouple message processing from the connection state (messages can be queued while we're connecting).
+    private val mailbox = Channel<Action>()
```

Comment:
This mailbox is never reset. It's not a problem because the matching between user requests and data sent to and received from electrum servers is done by jobs that are specific to an electrum server connection and will always be consistent, correct ? 
Also, mailbox is a "rendez-vous" channel so `send()` will suspend until `received()` is called which means that users can safely call the electrum client even while it is re-connecting to a different server ?

Reply:
Thanks, this is indeed one of the tricky parts: I added a larger comment in https://github.com/ACINQ/lightning-kmp/pull/512/commits/b2f72d31ce6be3b5a97973e2cbc9f71a5b5c7a7c to explain this more.